### ISSUE TITLE
Out of memory while editing

### ISSUE THREAD

**Issue by huguesv** [Microsoft/PTVS#499]
VS 2015 RC
PTVS Dev 2015-06-19 VS 2015

I had been using that VS session for maybe 2 hours.  I was in the process of adding a decorator to every test in blobservice (file is about 4000 lines long).  In a span of about 60 seconds, it started getting very slow, then unresponsive.  I saved dump file in our bug share under 'hugues-outofmemory', it's 3600MB.

**Comment by huguesv** [comment]
I've added the whole project to that folder.  Unzip that and open the solution.
Open `test_storage_blob.py` in the editor.
Scroll down to line 455.
Press CTRL-C to copy that whole line to your clipboard.  It's the `@record` line.
Scroll to the next test (line 470) and CTLR-V to add the decorator to the test.
Repeat this for every test.  Scroll down using the down arrow key.
I checked the memory before I started, and it was around 300MB.  I wasn't even done with half the file, and memory was at 1000MB.

**Comment by zooba** [comment]
So it's fairly complicated as to how this comes about, but here's my best written description that doesn't involve a whiteboard:

We have two modules:

A.py
```python
from B import h
def f(x): return h(x)

f(1)
```

B.py
```python
def g(x): return None
def h(x): return g(x)
```

On the first pass, we create a `FunctionAnalysisUnit` (henceforth FAU) for each function assuming no types for the parameter (`f`, `g`, `h`). When we analyze the `f(1)` we create a second FAU for `f(int)`. If there were another call to `f`, we would create a third FAU because the _call chain_ is different.

(The call chain is the sequence of `CallExpression` nodes leading to the call. This is what lets us get analysis for parameter sets uniquely.)

We also add a dependency to the return value of `f(int)`, so that whenever it updates we will reanalyze this function (which in this case does not matter, but if we were assigning the result we'd need to do this). The dependency is a set on `f(int)` that now contains a reference to our module's analysis unit.

When we analyze the `f(int)` FAU we will create a `h(int)` FAU for the `h` function. Because the call has come from `A`, we will key it off `A`'s version to ensure we drop the FAU when `A` is reanalyzed. The call chain for `h(int)` is `f(1)->h(x)`, which gives us a unique analysis for the original call to `f`. We also add a dependency from `f(1)->h(x)`'s return value, which is a reference in a set pointing to `h(int)`'s FAU.

Now when we analyze the `h(int)` FAU we will create a `g(int)` FAU with call chain `f(1)->h(x)->g(x)`. (Call chains are typically capped at three long, so calls from `g` are only distinguished as far back as the `h(x)`, so `x` is now the union of all calls to `f`). _This_ call has come from `B`, so we key it off `B`'s version and add the return value dependency.

Here's where things get tricky: we reanalyze `A` again without modifying it (or with an innocuous modification).

The entire process above repeats, but at any point where we keyed off `A`, we will notice that the version has changed and drop the old references. So `h` no longer has a reference to the old `h(int)` and will create a new one with the new (unchanged) type of `x`.

_However_, two things go wrong when we get to the call to `g`.

First: the FAU was not keyed off of `A`, but `B` whose version has not changed. So we don't drop the old FAU which keeps the _old_ `h(int)` FAU alive because it is still a dependency of the return value.

Second: we don't recognize that `f(1)->h(x)->g(x)` is the same call chain because we do reference comparisons on the `CallExpression` nodes, and reparsing `A` has changed the identity of the first two nodes. So we add a new FAU without dropping the old one.

Multiply this by 100 calls to `h` from `A` and by 100 edits and you have thousands of FAUs being "leaked". As soon as you reanalyze `B` all the unnecessary FAUs are dropped and memory usage returns to normal.

I don't see any easy way to fix this for 2.2. The extra FAUs are correctly left behind (as in, they are indistinguishable from ones that should be removed), we have no good concept of value comparisons of AST nodes, and no easy way to detect or clear the dependencies. I'm open to suggestions.

**Comment by int19h** [comment]
Is this a regression compared to 2.1?

**Comment by zooba** [comment]
I think Dino fixed this, but we need to check.

**Comment by huguesv** [comment]
It's not fixed, but at least now it's in the analyzer process, not devenv.exe, so there's more memory to spare.

Analyzer process is the blue line. This is with the project mentioned in my original comments.

![image](https://user-images.githubusercontent.com/1696845/31508670-ac31cf22-af33-11e7-89cc-c8c4a3adacfd.png)

**Comment by zooba** [comment]
So it turns out that most of this issue had in fact been resolved, and there was just one lingering reference leak that's being resolved in #3332.

However, in the code from #3298 we still see memory growing with each reanalysis, and this time it seems to be due to nested functions being assigned to dynamically generated classes. I'll investigate that over on #3298.
